frontend-v2/src/features/trial/DoseRow.tsx

```diff
@@ -0,0 +1,194 @@
+import { Delete } from "@mui/icons-material";
+import { IconButton, TableCell, TableRow, Typography } from "@mui/material";
+import { FC, useEffect, useMemo } from "react";
+import FloatField from "../../components/FloatField";
+import IntegerField from "../../components/IntegerField";
+import UnitField from "../../components/UnitField";
+import {
+  DoseRead,
+  Protocol,
+  UnitRead,
+  useDoseRetrieveQuery,
+  useDoseUpdateMutation,
+} from "../../app/backendApi";
+import { Control, useForm } from "react-hook-form";
+import useDirty from "../../hooks/useDirty";
+import useInterval from "../../hooks/useInterval";
+
+type Props = {
+  baseUnit?: UnitRead;
+  control: Control<Protocol, any, Protocol>;
+  disabled: boolean;
+  doseId: number;
+  index: number;
+  isPreclinical: boolean;
+  minStartTime: number;
+  removeDose: (index: number) => void;
+  selectedAmountLabel: string;
+  timeUnit?: UnitRead;
+};
+
+const DoseRow: FC<Props> = ({
+  baseUnit,
+  control,
+  disabled,
+  doseId,
+  index,
+  isPreclinical,
+  minStartTime,
+  removeDose,
+  selectedAmountLabel,
+  timeUnit,
+}) => {
+  const { data: dose, refetch: refetchDose } = useDoseRetrieveQuery(
+    { id: doseId },
+    { skip: !doseId },
+  );
+  const {
+    control: doseControl,
+    handleSubmit,
+    formState: { isDirty },
+    reset,
+  } = useForm<DoseRead>({
+    defaultValues: dose,
+  });
+  useDirty(isDirty);
+
+  useEffect(() => {
+    reset(dose);
+  }, [dose, reset]);
+
+  const [updateDose] = useDoseUpdateMutation();
+
+  const handleSave = useMemo(
+    () =>
+      handleSubmit(async (data) => {
+        if (JSON.stringify(data) !== JSON.stringify(dose)) {
+          await updateDose({ id: doseId, dose: data });
+          refetchDose();
```

Comment:
A callback here, to refetch the protocol in the parent component, might fix the 'Add new row' button.